Fix __call__ for Bidirectional wrapper

myutwo150 opened the pull request:
Fix a bug in model topology when `initial_state` is provided to the `Bidirectional` wrapper. `initial_state` should be handled in a same way as in `RNN.__call__`.

```python
input1 = Input(shape=(None, 5))
input2 = Input(shape=(None, 5))
states = Bidirectional(GRU(2, return_state=True))(input1)[1:]
out = Bidirectional(GRU(2, return_sequences=True))(input2, initial_state=states)
model = Model([input1, input2], out)
print(model.layers)
model.summary()
```
Outputs:
```
[<keras.engine.topology.InputLayer object at 0x117d760f0>, <keras.layers.wrappers.Bidirectional object at 0x117ec5550>]
_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================
input_2 (InputLayer)         (None, None, 5)           0
_________________________________________________________________
bidirectional_2 (Bidirection (None, None, 4)           96
=================================================================
Total params: 96
Trainable params: 96
Non-trainable params: 0
_________________________________________________________________

```
This PR basically copies `RNN.__call__` into `Bidirectional.__call__`, but without the lines handling `constants`.

Also modifies `wrappers_test.test_Bidirectional_state_reuse` to cover this use case.

fchollet commented on keras/layers/wrappers.py:
```diff
@@ -313,6 +314,51 @@ def compute_output_shape(self, input_shape):
+        self.forward_layer.state_spec = additional_specs[:num_states // 2]
+        self.backward_layer.state_spec = additional_specs[num_states // 2:]
+
+        is_keras_tensor = hasattr(additional_inputs[0], '_keras_history')
```
There is a backend method to do this. Do not rely on private attributes

fchollet commented on keras/layers/wrappers.py:
```diff
@@ -313,6 +314,51 @@ def compute_output_shape(self, input_shape):
+
+        is_keras_tensor = hasattr(additional_inputs[0], '_keras_history')
+        for tensor in additional_inputs:
+            if hasattr(tensor, '_keras_history') != is_keras_tensor:
```
Same

fchollet commented on keras/layers/wrappers.py:
```diff
@@ -313,6 +314,51 @@ def compute_output_shape(self, input_shape):
+            if hasattr(tensor, '_keras_history') != is_keras_tensor:
+                raise ValueError('The initial state of a Bidirectional'
+                                 ' layer cannot be specified with a mix of'
+                                 ' Keras tensors and non-Keras tensors')
```
Clarify the error message; a "Keras tensor" is the output of a Keras layer

myutwo150 commented:
Fixed issues mentioned in the above comments. The corresponding lines in `recurrent.py` are also changed.

Also replaced the call `self.forward_layer._standardize_args` with equivalent lines since it leads to an error when used with `ConvLSTM2D`.

Youmna-H commented:
This fix also needs to added to TimeDistributed layer to enable using initial_state
